### ISSUE TITLE
Large manifest file causes deb-s3 not to work

### ISSUE THREAD

**Issue by jjshoe** [krobertson/deb-s3#72]
I've got a 60MB Packages file that prevents deb-s3 from completing. We upload upload upload and never delete packages. I'm currently investigating more and will update this ticket with more details as I dig.

**Comment by jjshoe** [comment]
Did some poking, found a couple of things, though nothing that was a smoking gun for me.

https://github.com/krobertson/deb-s3/blob/master/lib/deb/s3/manifest.rb#L82 - It looks like to me that this is getting the results of the package.erb and cramming them all into an array, when it's done, it then calls a join. This was causing my desktop to kill the process for running out of memory. The way to do this would be to write the contents out to the file as generated. I played with this some, and while it prevented it from getting killed on my desktop, it didn't help with overall speed.

All the time was sucked in that generate method though. It seems like the correct approach would be:
1) See if I need to change an existing package
2) Write line by line until that existing package
3) Template
4) Write line by line until the end

If the package didn't exist it could just be tacked onto the end.

Another performance piece would be instead of writing out manifest into the FH for compressing you could just copy or compress Packages.

I ran it through ruby-prof, however github won't let me attach anything other then images, so you can grab the output here: https://www.dropbox.com/s/pl5kvkounilzvfl/stat.tar?dl=0

**Comment by jjshoe** [comment]
I found the root cause, my Packages file had lots of entrys of just periods:

```
[jjshoe@joel-desktop Downloads]$ du -hs Packages 
121M	Packages
[jjshoe@joel-desktop Downloads]$ grep -ve '^\s*.\s*$' Packages > test
[jjshoe@joel-desktop Downloads]$ du -hs test 
88K	test
[jjshoe@joel-desktop Downloads]$
```

**Comment by jjshoe** [comment]
Note: going to hold the ticket open for a bit, see if I can figure out what caused this.

**Comment by leedm777** [comment]
I've looked into this a bit. It appears that blank continuation lines in a description (which are rendered as ` .` in the Packages file) are duplicated when it goes through the parsing process. So for every successive upload, the number of blank lines are doubled.

**Comment by jjshoe** [comment]
@leedm777 - perhaps: https://github.com/krobertson/deb-s3/blob/master/lib/deb/s3/manifest.rb#L46

**Comment by leedm777** [comment]
The massive number of dots we were seeing were caused by a parsing bug. PR posted, which even includes a test :-)

**Comment by leedm777** [comment]
FWIW, it looks like this bug was introduced between 0.6.2 and 0.7.0.

**Comment by JeanFred** [comment]
I am having the same problem − dot lines get duplicated, resulting in a huge Packages file eventually leading deb-s3 to crash when manipulating it.

**Comment by leedm777** [comment]
@JeanFred Can you try out [my patch](https://github.com/krobertson/deb-s3/issues/74) to see if it works for you?

I also wrote a [fixup command](https://github.com/leedm777/deb-s3/tree/fixup) in a branch on my work. I used it to get our Packages files back in order. It's a quick hack, so I highly recommend backups before trying it.

**Comment by JeanFred** [comment]
@leedm777 I have tried your patch and I’m happy to report that it does work for me − thanks! I tried uploading packages with deb-s3 − which duplicated dots − and then your version − which did not.

I also used (after backup ;-) your fixup command, which worked like a charm and basically saved our lives. Thanks *a lot* for this.

**Comment by krobertson** [comment]
Merged the PR and released 0.7.1 earlier this morning. Thanks!

**Comment by JeanFred** [comment]
Thanks @krobertson and @leedm777 !
